addon/mixins/ember-hooks/index.ts

```diff
@@ -0,0 +1,126 @@
+import Mixin from '@ember/object/mixin';
+import Component from '@ember/component';
+import { inject as service } from '@ember/service';
+import { depsAreEqual } from './util';
+
+type MemoizedStateEntry = [any, any[] | null];
+
+export interface IHookStore {
+  memoizedState: {
+    [key: number]: MemoizedStateEntry
+  }
+}
+
+export interface IEmberHooksComponent extends Component {
+  hooks(instance: Component): void,
+  instanceProxy: ProxyHandler<any>,
+  _state: string,
+  _hookStore: IHookStore,
+}
+
+let hookCallIndex = 0;
+let currentInstance: IEmberHooksComponent | null = null;
+
+const EmberHooksMixin = Mixin.create({
+  init(this: IEmberHooksComponent) {
+    this._super(...arguments);
+    const hooks = this.get('hooks');
+    if (hooks) {
+      this.set('_hookStore', { memoizedState: {} });
+      hookCallIndex = 0;
+      currentInstance = this;
+      const vals = hooks.call(this);
+      this.setProperties(vals);
+      currentInstance = null;
+    }
+  },
+  willRender(this: IEmberHooksComponent) {
+    this._super(...arguments);
+    const hooks = this.get('hooks');
+    if (hooks) {
+      hookCallIndex = 0;
+      currentInstance = this;
+      const vals = hooks.call(currentInstance);
+      currentInstance.setProperties(vals);
+      currentInstance = null;
+    }
+  },
+});
+
+export const useProperties = defaultProps => {
+  const self = currentInstance;
+  if (self === null) {
+    throw new Error('Unable to find Ember instance');
+  }
+
+  if (self._state === 'preRender' && !self.instanceProxy) {
+    self.setProperties(defaultProps);
+    const observed = self.getProperties(Object.keys(defaultProps));
+
+    self.instanceProxy = new Proxy(observed, {
+      get: function(target, prop, receiver) {
+        return self.get(prop);
+      },
+      set(obj, prop, value) {
+        obj[prop] = value;
+        return self.set(prop, value);
+      },
+    });
+
+  }
+
+  return self.instanceProxy;
+};
+
+export const useStore = () => {
+  const self = currentInstance;
+  if (self === null) {
+    throw new Error('Unable to find Ember instance');
+  }
+
+  if (self._state === 'preRender') self.set('store', service('store'));
+  return self.get('store');
+};
+
+export const withHooks = (...args) => {
+  const config = args.pop(args.length - 1);
+  return Component.extend(EmberHooksMixin, ...args, {
+    hooks() {
+      return config(this.attrs);
+    },
+  });
+}
+
+/**
+ * useMemo is a hook that will only recompute a potentially expensive
+ * value when one of it's dependencies change.
+ *
+ * @param {function} createMemoizedValue - function to compute an expensive value
+ * @param {Array} deps - Array of dependencies
+ */
+export const useMemo = <T>(createMemoizedValue: () => T, deps?: any[] | null): T => {
+  const instance = currentInstance;
+  if (instance === null) {
```

Comment:
Should we abstract the instance null check? Seems like it will be consistent across hooks

Reply:
I tried abstracting it, but I couldn't get TS to not complain about the typeguard when in a separate method (It would still complain about instance potentially being null). I'll give it a second look though.

Reply:
the hooks function itself could check the instance before running. No point running any of them if the instance is null

Reply:
@twaite Yes but then TS will not work correctly. TS has not way of ensuring that the currentInstance was not set to null in between calls to a hook.

Reply:
Only other way would be an explicit cast of currentInstance in the hooks. 

```js
const instance = (currentInstance as IEmberHooksComponent);
```